File: ROS/src/beluga/beluga/beluga_serial.py
from io import TextIOWrapper
import sys
import os
import signal
import serial
import serial.tools.list_ports as list_ports
from typing import List, Dict, Optional, Union, TextIO, Callable, Any, Tuple
import multiprocessing as mp
import multiprocessing.queues as mp_queues
import queue
import time
import itertools
from beluga.beluga_frame import BelugaFrame, FrameType

TARGETS = [
    'CMU Beluga',
    'SEGGER J-Link'
]


class BelugaEntryError(Exception):
    def __init__(self, msg: str, header: bool = False):
        self.header = header
        super().__init__(msg)


class BelugaNeighbor:
    def __init__(self, neighbor: dict):
        self._id: int = neighbor["ID"]
        self._range: float = 0.0
        self._rssi: int = 0
        self._time: int = 0
        self._exchange_id: int = 0
        self._updated = False
        self.update(neighbor)

    def __iter__(self):
        return iter([("RANGE", self._range), ("RSSI", self._rssi), ("TIMESTAMP", self._time), ("EXCHANGE", self._exchange_id)])

    @property
    def id(self) -> int:
        return self._id

    @property
    def range(self) -> float:
        return self._range

    @property
    def rssi(self) -> int:
        return self._rssi

    @property
    def time(self) -> int:
        return self._time

    @property
    def exchange(self) -> int:
        return self._exchange_id

    @property
    def updated(self) -> bool:
        return self._updated

    @updated.setter
    def updated(self, update: bool):
        self._updated = update

    def update(self, neighbor: dict):
        self._range = neighbor["RANGE"]
        self._rssi = neighbor["RSSI"]
        self._time = neighbor["TIMESTAMP"]
        if "EXCHANGE" in neighbor.keys():
            self._exchange_id = neighbor["EXCHANGE"]
        self._updated = True


class BelugaNeighborList:
    def __init__(self):
        self._list: Dict[int, BelugaNeighbor] = {}
        self._neighbors_update: bool = False
        self._range_update: bool = False
        return

    def update(self, updates: List[Dict[str, Union[int, float]]]):
        for entry in updates:
            if entry["ID"] not in self._list.keys():
                self._list[entry["ID"]] = BelugaNeighbor(entry)
                self._range_update = True
                self._neighbors_update = True
            else:
                self._list[entry["ID"]].update(entry)
                self._range_update = True

    def remove_neighbor(self, neighbor_id: int):
        if neighbor_id in self._list.keys():
            del self._list[neighbor_id]
            self._neighbors_update = True

    def get_updates(self) -> Dict[int, Dict[str, Union[int, float]]]:
        ret = {}
        for x in self._list.values():
            if x.updated:
                ret[x.id] = dict(x)
                x.updated = False
        self._range_update = False
        return ret

    def get_neighbors(self) -> Dict[int, Dict[str, Union[int, float]]]:
        ret = {}
        for x in self._list.values():
            ret[x.id] = dict(x)
        self._neighbors_update = False
        return ret

    def clear(self):
        if self._list:
            self._list.clear()
            self._neighbors_update = True
            self._range_update = False

    @property
    def neighbor_update(self) -> bool:
        return self._neighbors_update

    @property
    def range_update(self) -> bool:
        return self._range_update



class BelugaQueue(mp_queues.Queue):
    def __init__(self, maxsize: int = 1, update_old_items: bool = True):
        self._update = update_old_items
        super().__init__(maxsize=maxsize, ctx=mp.get_context())

    def put(self, obj, block: bool = True, timeout: Optional[float] = None) -> None:
        try:
            super().put(obj, block, timeout)
        except queue.Full:
            try:
                item = self.get_nowait()
            except queue.Empty:
                # Queue item got removed between the put and the get operations. Place new item
                pass
            else:
                if self._update:
                    # This will update the existing keys, add new key-value pairs if not present, and keep keys that
                    # are not present in the new obj unchanged
                    item.update(obj)
                    obj = item
                # else drop the old item...
            super().put(obj, block, timeout)
        return

    def clear(self):
        for _ in itertools.repeat(None, self.qsize()):
            try:
                self.get_nowait()
            except queue.Empty:
                # Queue is now empty
                break


class BelugaSerial:
    # TODO: Discuss how we want to handle case where USB communication fails. Do we want to log the error and try to reconnect? Do we want to terminate the node?
    def __init__(self,
                 baud: int = 115200,
                 timeout: float = 2.0,
                 serial_timeout: float = 0.1,
                 max_lines_read: int = 16,
                 port: Optional[str] = None,
                 neighbor_update_func: Optional[Callable[[dict], None]] = None,
                 range_update_func: Optional[Callable[[dict], None]] = None,
                 range_event: Optional[Callable[[dict], None]] = None,
                 logger_func: Optional[Callable[[Any], None]] = None):

        self._logger = logger_func

        if port is None:
            targets = self._find_ports(TARGETS)
            self._serial = None
            if not targets:
                raise FileNotFoundError(f'Unable to find a given target. Valid targets: {TARGETS}')
            for target in TARGETS:
                if target in targets.keys():
                    for port in targets[target]:
                        try:
                            self._log(f"Trying to connect to {target}: {port}")
                            self._serial = serial.Serial(port=port, baudrate=baud, timeout=serial_timeout,
                                                         exclusive=True)
                            self._log(f'Connected to {target}: {port}')
                            break
                        except serial.SerialException:
                            self._log(f'{port} is busy')
                            pass
                    break
            if self._serial is None:
                raise FileNotFoundError(f'Unable to find a given target. Valid targets: {TARGETS}')
        else:
            self._serial = serial.Serial(port=port, baudrate=baud, timeout=serial_timeout, exclusive=True)

        self._read_max_lines: int = max_lines_read
        self._timeout: float = timeout
        self._neighbors = BelugaNeighborList()

        self._rx_task: Optional[mp.Process] = None
        self._batch_queue: BelugaQueue = BelugaQueue(10, False)

        self._processing_task: Optional[mp.Process] = None
        self._response_q: BelugaQueue = BelugaQueue(update_old_items=False)
        self._command_sent: mp.Event = mp.Event()
        # Used to block the reboot command until done rebooting
        self._reboot_done: mp.Event = mp.Event()

        if neighbor_update_func is None:
            self._neighbors_queue: Optional[BelugaQueue] = BelugaQueue()
            self._neighbors_callback = None
        else:
            self._neighbors_queue = None
            self._neighbors_callback = neighbor_update_func

        if range_update_func is None:
            self._ranges_queue: Optional[BelugaQueue] = BelugaQueue()
            self._ranges_update_callback = None
        else:
            self._ranges_queue = None
            self._ranges_update_callback = range_update_func

        if range_event is None:
            self._range_event_queue: Optional[BelugaQueue] = BelugaQueue()
            self._range_event_callback = None
        else:
            self._range_event_queue = None
            self._range_event_callback = range_event


    def _log(self, s):
        if self._logger is not None:
            self._logger(s)

    def _publish_neighbor_update(self):
        if self._neighbors.neighbor_update:
            if self._neighbors_queue is not None:
                self._neighbors_queue.put(self._neighbors.get_neighbors(), block=False)
            else:
                self._neighbors_callback(self._neighbors.get_neighbors())

    def _publish_range_update(self):
        if self._neighbors.range_update:
            if self._ranges_queue is not None:
                self._ranges_queue.put(self._neighbors.get_updates(), block=False)
            else:
                self._ranges_update_callback(self._neighbors.get_updates())

    def _publish_range_event(self, event):
        if self._range_event_queue is not None:
            self._range_event_queue.put(event, block=False)
        else:
            self._range_event_callback(event)

    def _publish_response(self, response):
        if self._command_sent.is_set():
            self._command_sent.clear()
            self._response_q.put(response)

    def _process_reboot(self, payload):
        if self._ranges_queue is not None:
            self._ranges_queue.clear()
        if self._neighbors_queue is not None:
            self._neighbors_queue.clear()
        if self._range_event_queue is not None:
            self._range_event_queue.clear()
        self._neighbors.clear()
        if self._reboot_done.is_set():
            # Unexpected reboot
            self._log("Beluga rebooted unexpectedly")
            os.kill(os.getppid(), signal.SIGUSR1)
        else:
            self._reboot_done.set()

    @staticmethod
    def _find_ports(targets: List[str]) -> Dict[str, List[str]]:
        ret: Dict[str, List[str]] = {}
        ports = list(list_ports.comports())

        for port in ports:
            dev_name = f'{port.manufacturer} {port.product}'
            if dev_name in targets:
                if dev_name in ret.keys():
                    ret[dev_name].append(port.device)
                else:
                    ret[dev_name] = [port.device]
        return ret

    def _process_frames(self):
        while True:
            frame: BelugaFrame = self._batch_queue.get()
            match frame.type:
                case FrameType.UPDATES:
                    self._neighbors.update(frame.payload)
                case FrameType.EVENT:
                    self._publish_range_event(frame.payload)
                case FrameType.DROP:
                    self._neighbors.remove_neighbor(frame.payload)
                case FrameType.RESPONSE:
                    self._publish_response(frame.payload)
                case FrameType.START:
                    self._process_reboot(frame.payload)
                case _:
                    self._log("Invalid frame type")
            self._publish_neighbor_update()
            self._publish_range_update()

    def _process_frames_entry(self):
        try:
            self._process_frames()
        except KeyboardInterrupt:
            pass # Shutting down

    def _process_rx_buffer(self, buf: bytearray) -> bytearray:
        frame_start, frame_size, _ = BelugaFrame.frame_present(buf)
        if frame_start < 0:
            return buf

        frame = BelugaFrame.extract_frame(buf, frame_start)
        self._batch_queue.put(frame, block=False)
        return buf[frame_start + frame_size:]

    def _read_serial(self):
        rx = b""
        while True:
            if self._serial.inWaiting() > 0:
                rx += self._serial.read(self._serial.inWaiting())
            rx = self._process_rx_buffer(rx)

    def _read_task_entry(self):
        try:
            self._read_serial()
        except KeyboardInterrupt:
            pass # Shutting down

    def _send_command(self, command: bytes) -> str:
        self._serial.write(command)
        try:
            self._command_sent.set()
            ret = self._response_q.get(timeout=self._timeout)
        except queue.Empty:
            ret = 'Response timed out'
        return ret

    def start_uwb(self) -> str:
        ret = self._send_command(b'AT+STARTUWB\r\n')
        return ret

    def stop_uwb(self) -> str:
        ret = self._send_command(b'AT+STOPUWB\r\n')
        return ret

    def start_ble(self) -> str:
        ret = self._send_command(b'AT+STARTBLE\r\n')
        return ret

    def stop_ble(self) -> str:
        ret = self._send_command(b'AT+STOPBLE\r\n')
        return ret

    def id(self, new_id: Optional[int] = None) -> str:
        if new_id is not None:
            command = f'AT+ID {new_id}\r\n'.encode()
        else:
            command = b'AT+ID\r\n'
        ret = self._send_command(command)
        return ret

    def bootmode(self, boot_mode: Optional[int] = None) -> str:
        if boot_mode is not None:
            command = f'AT+BOOTMODE {int(boot_mode)}\r\n'.encode()
        else:
            command = b'AT+BOOTMODE\r\n'
        ret = self._send_command(command)
        return ret

    def rate(self, rate: Optional[int] = None) -> str:
        if rate is not None:
            command = f'AT+RATE {rate}\r\n'.encode()
        else:
            command = b'AT+RATE\r\n'
        ret = self._send_command(command)
        return ret

    def channel(self, channel: Optional[int] = None) -> str:
        if channel is not None:
            command = f'AT+CHANNEL {channel}\r\n'.encode()
        else:
            command = b'AT+CHANNEL\r\n'
        ret = self._send_command(command)
        return ret

    def reset(self) -> str:
        ret = self._send_command(b'AT+RESET\r\n')
        return ret

    def timeout(self, timeout: Optional[int] = None) -> str:
        if timeout is not None:
            command = f'AT+TIMEOUT {timeout}\r\n'.encode()
        else:
            command = b'AT+TIMEOUT\r\n'
        ret = self._send_command(command)
        return ret

    def tx_power(self, max_power: Optional[int] = None) -> str:
        if max_power is not None:
            command = f'AT+TXPOWER {max_power}\r\n'.encode()
        else:
            command = b'AT+TXPOWER\r\n'
        ret = self._send_command(command)
        return ret

    def stream_mode(self, updates_only: Optional[int] = None) -> str:
        if updates_only is not None:
            command = f'AT+STREAMMODE {updates_only}\r\n'.encode()
        else:
            command = b'AT+STREAMMODE\r\n'
        ret = self._send_command(command)
        return ret

    def twr_mode(self, mode: Optional[int] = None) -> str:
        if mode is not None:
            command = f'AT+TWRMODE {mode}\r\n'.encode()
        else:
            command = b'AT+TWRMODE\r\n'
        ret = self._send_command(command)
        return ret

    def led_mode(self, led_mode: Optional[int] = None) -> str:
        if led_mode is not None:
            command = f'AT+LEDMODE {led_mode}\r\n'.encode()
        else:
            command = b'AT+LEDMODE\r\n'
        ret = self._send_command(command)
        return ret

    def reboot(self) -> str:
        self._reboot_done.clear()
        ret = self._send_command(b'AT+REBOOT\r\n')
        self._reboot_done.wait()
        return ret

    def pwr_amp(self, enable_pwr_amp: Optional[int] = None) -> str:
        if enable_pwr_amp is not None:
            command = f'AT+PWRAMP {enable_pwr_amp}\r\n'.encode()
        else:
            command = b'AT+PWRAMP\r\n'
        ret = self._send_command(command)
        return ret

    def antenna(self, antenna: Optional[int] = None) -> str:
        if antenna is not None:
            command = f'AT+ANTENNA {antenna}\r\n'.encode()
        else:
            command = b'AT+ANTENNA\r\n'
        ret = self._send_command(command)
        return ret

    def time(self) -> str:
        command = b'AT+TIME\r\n'
        ret = self._send_command(command)
        return ret

    def format(self, json_format: Optional[int] = None) -> str:
        if json_format is not None:
            command = f'AT+FORMAT {json_format}\r\n'.encode()
        else:
            command = b'AT+FORMAT\r\n'
        ret = self._send_command(command)
        return ret

    def deepsleep(self) -> str:
        command = b'AT+DEEPSLEEP\r\n'
        ret = self._send_command(command)
        return ret

    def datarate(self, rate: Optional[int] = None) -> str:
        if rate is not None:
            command = f'AT+DATARATE {rate}\r\n'.encode()
        else:
            command = b'AT+DATARATE\r\n'
        ret = self._send_command(command)
        return ret

    def preamble(self, length: Optional[int] = None) -> str:
        if length is not None:
            command = f'AT+PREAMBLE {length}\r\n'.encode()
        else:
            command = b'AT+PREAMBLE\r\n'
        ret = self._send_command(command)
        return ret

    def pulserate(self, rate: Optional[int] = None) -> str:
        if rate is not None:
            command = f'AT+PULSERATE {rate}\r\n'.encode()
        else:
            command = b'AT+PULSERATE\r\n'
        ret = self._send_command(command)
        return ret

    def start(self):
        if self._rx_task is not None or self._processing_task is not None:
            raise RuntimeError('Please stop before restarting')
        self._processing_task = mp.Process(target=self._process_frames_entry)
        self._rx_task = mp.Process(target=self._read_task_entry)

        self._processing_task.start()
        self._rx_task.start()

    def stop(self):
        if self._rx_task is not None:
            self._rx_task.kill()
            while self._rx_task.is_alive():
                time.sleep(0.1)
            self._rx_task.close()
        if self._processing_task is not None:
            self._processing_task.kill()
            while self._processing_task.is_alive():
                time.sleep(0.1)
            self._processing_task.close()
        return

    def close(self):
        if self._serial.isOpen():
            self._serial.close()

    def get_neighbors(self) -> Tuple[bool, Dict[int, Dict[str, Union[int, float]]]]:
        # Needed to indicate if the queue is just empty or if the only neighbor got removed
        if self._neighbors_queue is None:
            raise ValueError("Reporting neighbor updates to a callback function")
        update = True
        ret = {}
        try:
            ret = self._neighbors_queue.get_nowait()
        except queue.Empty:
            update = False
        return update, ret

    def get_ranges(self) -> Dict[int, Dict[str, Union[int, float]]]:
        if self._ranges_queue is None:
            raise ValueError("Reporting range updates to a callback function")
        ret = {}
        try:
            ret = self._ranges_queue.get_nowait()
        except queue.Empty:
            pass
        return ret

    def get_range_event(self) -> Dict[int, Dict[str, int]]:
        if self._range_event_queue is None:
            raise ValueError("Reporting range events to a callback function")
        ret = {}
        try:
            ret = self._range_event_queue.get_nowait()
        except queue.Empty:
            pass
        return ret


def main():
    import datetime as dt
    beluga = BelugaSerial()
    beluga.start()
    beluga.reboot()
    beluga.stream_mode(True)
    last_tr = dt.datetime.now()

    try:
        while True:
            ranges = beluga.get_ranges()
            if ranges:
                print(f'Ranges {ranges}')
            update, neighbors = beluga.get_neighbors()
            if update:
                print(f'Neighbors: {neighbors}')

            elapsed = dt.datetime.now() - last_tr
            if elapsed.microseconds > 100000:
                ret = beluga.time()
                print(ret)
                last_tr = dt.datetime.now()
    except KeyboardInterrupt:
        beluga.stop()
        beluga.close()


if __name__ == '__main__':
    main()

File: ROS/src/beluga/src/beluga_serial.cpp
/**
 * @file beluga_serial.cpp
 *
 * @brief
 *
 * @date 1/30/25
 *
 * @author tom
 */

#include <beluga/beluga_serial.hpp>
#include <chrono>
#include <filesystem>
#include <serial/serial.hpp>
#include <serial/tools/list_ports.hpp>
#include <stdexcept>
#include <thread>
#include <utility>

using namespace std::chrono_literals;
namespace fs = std::filesystem;

namespace BelugaSerial {

static const std::vector<target_pair> TARGETS = {{"CMU", "Beluga"},
                                                 {"SEGGER", "J-Link"}};

FileNotFoundError::FileNotFoundError(const char *msg) { _msg = msg; }

const char *FileNotFoundError::what() const noexcept { return _msg.c_str(); }

BelugaSerial::BelugaSerial() { _initialize(); }

BelugaSerial::BelugaSerial(
    const std::string &port, BaudRate baud,
    const std::chrono::milliseconds &timeout,
    const std::chrono::milliseconds &serial_timeout, uint32_t max_lines_read,
    std::function<void(const std::vector<BelugaNeighbor> &)> neighbor_update_cb,
    std::function<void(const std::vector<BelugaNeighbor> &)> range_updates_cb,
    std::function<void(const RangeEvent &)> range_event_cb,
    std::function<int(const char *, va_list)> logger_cb) {
    _initialize(baud, timeout, serial_timeout, max_lines_read, port,
                std::move(neighbor_update_cb), std::move(range_updates_cb),
                std::move(range_event_cb), std::move(logger_cb));
}

void BelugaSerial::_initialize(
    BaudRate baud, const std::chrono::milliseconds &timeout,
    const std::chrono::milliseconds &serial_timeout, uint32_t max_lines_read,
    const std::string &port,
    std::function<void(const std::vector<BelugaNeighbor> &)> neighbor_update_cb,
    std::function<void(const std::vector<BelugaNeighbor> &)> range_updates_cb,
    std::function<void(const RangeEvent &)> range_event_cb,
    std::function<int(const char *, va_list)> logger_cb) {
    _logger_cb = std::move(logger_cb);

    _serial.baudrate(baud);
    _serial.timeout(serial_timeout);
    _serial.exclusive(true);

    if (port.empty()) {
        std::map<target_pair, std::vector<std::string>> avail_ports;
        _find_ports(TARGETS, avail_ports);
        if (avail_ports.empty()) {
            throw FileNotFoundError("Unable to find a given target");
        }

        for (const auto &target : TARGETS) {
            if (avail_ports.find(target) == avail_ports.end()) {
                continue;
            }
            for (const auto &port_ : avail_ports[target]) {
                std::stringstream oss;
                try {
                    oss << "Trying to connect to " << target.first << " "
                        << target.second << ": " << port_ << "\n";
                    _log(oss.str().c_str());
                    _serial.port(port_);
                    _serial.open();
                    break;
                } catch (const Serial::SerialException &exc) {
                    oss.clear();
                    oss << exc.what() << "\n";
                    _log(oss.str().c_str());
                }
            }
        }

        if (!_serial.is_open()) {
            throw FileNotFoundError("Unable to open a target");
        }
    } else {
        _serial.port(port);
        _serial.open();
    }

    _read_max_lines = max_lines_read;
    _timeout = timeout;

    _neighbor_cb = std::move(neighbor_update_cb);
    _range_cb = std::move(range_updates_cb);
    _range_event_cb = std::move(range_event_cb);
}

BelugaSerial::~BelugaSerial() { this->close(); }

void BelugaSerial::_log(const char *msg, ...) {
    if (_logger_cb != nullptr) {
        va_list args;
        va_start(args, msg);
        _logger_cb(msg, args);
        va_end(args);
    }
}

void BelugaSerial::_publish_neighbor_update() {
    if (_neighbors.neighbor_updates()) {
        std::vector<BelugaNeighbor> updates;
        _neighbors.get_neighbors(updates);
        if (_neighbor_cb != nullptr) {
            _neighbor_cb(updates);
        } else {
            _neighbor_queue.put(updates, false);
        }
    }
}

void BelugaSerial::_publish_range_update() {
    if (_neighbors.range_updates()) {
        std::vector<BelugaNeighbor> updates;
        _neighbors.get_updates(updates);
        if (_range_cb != nullptr) {
            _range_cb(updates);
        } else {
            _range_queue.put(updates, false);
        }
    }
}

void BelugaSerial::_publish_range_event(RangeEvent event) {
    if (_range_event_cb != nullptr) {
        _range_event_cb(event);
    } else {
        _range_event_queue.put(event, false);
    }
}

void BelugaSerial::_publish_response(std::string &response) {
    if (_command_sent.is_set()) {
        _command_sent.clear();
        _response_queue.put(response);
    }
}

void BelugaSerial::_process_reboot(const std::string &) {
    _range_queue.clear();
    _neighbor_queue.clear();
    _range_event_queue.clear();
    _neighbors.clear();
    if (_reboot_done.is_set()) {
        _log("Beluga rebooted unexpectedly");
        if (_time_resync != nullptr) {
            std::thread t(_time_resync);
            t.detach();
        }
    } else {
        _reboot_done.set();
    }
}

void BelugaSerial::_find_ports(
    const std::vector<target_pair> &valid,
    std::map<target_pair, std::vector<std::string>> &avail_ports) {
    std::vector<SerialTools::SysFS> ports = SerialTools::comports();

    avail_ports.clear();

    for (const auto &port : ports) {
        target_pair target = {port.manufacturer(), port.product()};
        for (const auto &valid_target : valid) {
            if (valid_target == target) {
                avail_ports[target].push_back(port.device());
                break;
            }
        }
    }
}

void BelugaSerial::_process_frames() {
    while (_tasks_running) {
        BelugaFrame::DecodedFrame frame = _batch_queue.get();

        switch (frame.type) {
        case BelugaFrame::BelugaFrameType::NEIGHBOR_UPDATE:
            _neighbors.update(
                std::get<std::vector<BelugaFrame::NeighborUpdate>>(
                    frame.payload));
            break;
        case BelugaFrame::BelugaFrameType::RANGING_EVENT:
            _publish_range_event(std::get<RangeEvent>(frame.payload));
            break;
        case BelugaFrame::BelugaFrameType::NEIGHBOR_DROP:
            _neighbors.remove(std::get<uint32_t>(frame.payload));
            break;
        case BelugaFrame::BelugaFrameType::COMMAND_RESPONSE:
            _publish_response(std::get<std::string>(frame.payload));
            break;
        case BelugaFrame::BelugaFrameType::START_EVENT:
            _process_reboot(std::get<std::string>(frame.payload));
            break;
        default:
            _log("Invalid frame type");
            break;
        }
        _publish_neighbor_update();
        _publish_range_update();
    }
}

void BelugaSerial::_process_rx_buffer(std::vector<uint8_t> &buf) {
    while (true) {
        auto [frame_start, frame_size, _] =
            BelugaFrame::frame_present((const char *)buf.data(), buf.size());
        if (frame_start < 0) {
            return;
        }

        BelugaFrame frame;
        frame.parse_frame(buf, frame_start);
        BelugaFrame::DecodedFrame frame_ = frame.get_parsed_data();
        _batch_queue.put(frame_, false);
        buf.erase(buf.begin() + frame_start,
                  buf.begin() + frame_start + frame_size);
    }
}

void BelugaSerial::_read_serial() {
    std::vector<uint8_t> rx;

    while (_tasks_running) {
        if (_serial.in_waiting() > 0) {
            std::vector<uint8_t> buf;
            _serial.read_all(buf);
            rx.insert(rx.end(), buf.begin(), buf.end());
        }
        _process_rx_buffer(rx);
    }
}

std::string BelugaSerial::_send_command(const std::string &command) {
    std::vector<uint8_t> tx_data(command.begin(), command.end());
    std::string response;

    _serial.write(tx_data);
    _command_sent.set();
    try {
        response = _response_queue.get(false, _timeout);
    } catch (const BelugaQueueException &exc) {
        if (exc.reason() == BelugaQueueException::QUEUE_EMPTY) {
            response = "Response timed out";
        } else {
            throw;
        }
    }
    return response;
}

std::string BelugaSerial::start_uwb() {
    return _send_command("AT+STARTUWB\r\n");
}

std::string BelugaSerial::stop_uwb() { return _send_command("AT+STOPUWB\r\n"); }

std::string BelugaSerial::start_ble() {
    return _send_command("AT+STARTBLE\r\n");
}

std::string BelugaSerial::stop_ble() { return _send_command("AT+STOPBLE\r\n"); }

std::string BelugaSerial::id(const std::string &id_) {
    std::stringstream oss;
    oss << "AT+ID " << id_ << "\r\n";
    return _send_command(oss.str());
}

std::string BelugaSerial::bootmode(const std::string &mode) {
    std::stringstream oss;
    oss << "AT+BOOTMODE " << mode << "\r\n";
    return _send_command(oss.str());
}

std::string BelugaSerial::rate(const std::string &rate_) {
    std::stringstream oss;
    oss << "AT+RATE " << rate_ << "\r\n";
    return _send_command(oss.str());
}

std::string BelugaSerial::channel(const std::string &channel_) {
    std::stringstream oss;
    oss << "AT+CHANNEL " << channel_ << "\r\n";
    return _send_command(oss.str());
}

std::string BelugaSerial::reset() { return _send_command("AT+RESET\r\n"); }

std::string BelugaSerial::timeout(const std::string &timeout_) {
    std::stringstream oss;
    oss << "AT+TIMEOUT " << timeout_ << "\r\n";
    return _send_command(oss.str());
}

std::string BelugaSerial::tx_power(const std::string &power) {
    std::stringstream oss;
    oss << "AT+TXPOWER " << power << "\r\n";
    return _send_command(oss.str());
}

std::string BelugaSerial::stream_mode(const std::string &updates_only) {
    std::stringstream oss;
    oss << "AT+STREAMMODE " << updates_only << "\r\n";
    return _send_command(oss.str());
}

std::string BelugaSerial::twr_mode(const std::string &mode) {
    std::stringstream oss;
    oss << "AT+TWRMODE " << mode << "\r\n";
    return _send_command(oss.str());
}

std::string BelugaSerial::led_mode(const std::string &mode) {
    std::stringstream oss;
    oss << "AT+LEDMODE " << mode << "\r\n";
    return _send_command(oss.str());
}

std::string BelugaSerial::reboot() {
    _reboot_done.clear();
    std::string ret = _send_command("AT+REBOOT\r\n");
    _reboot_done.wait();
    return ret;
}

std::string BelugaSerial::pwr_amp(const std::string &mode) {
    std::stringstream oss;
    oss << "AT+PWRAMP " << mode << "\r\n";
    return _send_command(oss.str());
}

std::string BelugaSerial::antenna(const std::string &antenna) {
    std::stringstream oss;
    oss << "AT+ANTENNA " << antenna << "\r\n";
    return _send_command(oss.str());
}

std::string BelugaSerial::time() { return _send_command("AT+TIME\r\n"); }

std::string BelugaSerial::format(const std::string &mode) {
    std::stringstream oss;
    oss << "AT+FORMAT " << mode << "\r\n";
    return _send_command(oss.str());
}

std::string BelugaSerial::deepsleep() {
    return _send_command("AT+DEEPSLEEP\r\n");
}

std::string BelugaSerial::datarate(const std::string &rate_) {
    std::stringstream oss;
    oss << "AT+DATARATE " << rate_ << "\r\n";
    return _send_command(oss.str());
}

std::string BelugaSerial::preamble(const std::string &preamble) {
    std::stringstream oss;
    oss << "AT+PREAMBLE " << preamble << "\r\n";
    return _send_command(oss.str());
}

std::string BelugaSerial::pulserate(const std::string &pr) {
    std::stringstream oss;
    oss << "AT+PULSERATE " << pr << "\r\n";
    return _send_command(oss.str());
}

void BelugaSerial::start() {
    if (_tasks_running) {
        throw std::runtime_error("Please stop before restarting");
    }

    _tasks_running = true;

    _processing_task.task =
        std::packaged_task<void()>([this] { _process_frames(); });
    _processing_task.thread = std::thread(std::move(_processing_task.task));
    _rx_task.task = std::packaged_task<void()>([this] { _read_serial(); });
    _rx_task.thread = std::thread(std::move(_rx_task.task));
}

void BelugaSerial::stop() {
    const int max_attempts = 10;
    if (!_tasks_running) {
        return;
    }
    auto rx_future = _rx_task.task.get_future();
    auto process_future = _processing_task.task.get_future();

    _tasks_running = false;

    while (true) {
        auto status = rx_future.wait_for(10ms);
        if (status == std::future_status::ready) {
            break;
        }
    }
    _rx_task.thread.join();

    int retries = 0;
    for (; retries < max_attempts; retries++) {
        int attempts = 0;
        for (; attempts < max_attempts; attempts++) {
            auto status = process_future.wait_for(10ms);
            if (status == std::future_status::ready) {
                break;
            }
        }

        if (attempts >= max_attempts) {
            // Processing task hung itself. Need to cut it from the tree.
            BelugaFrame::DecodedFrame frame;
            frame.type = BelugaFrame::BelugaFrameType::NO_TYPE;
            _batch_queue.put(frame);
        } else {
            _processing_task.thread.join();
            break;
        }
    }

    if (retries >= max_attempts) {
        throw std::runtime_error("Processing task is still hanging...");
    }
}

void BelugaSerial::close() {
    stop();
    _serial.close();
}

bool BelugaSerial::get_neighbors(std::vector<BelugaNeighbor> &list) {
    bool update = true;
    list.clear();

    try {
        list = _neighbor_queue.get(false);
    } catch (const BelugaQueueException &exc) {
        if (exc.reason() == BelugaQueueException::QUEUE_EMPTY) {
            update = false;
        } else {
            throw;
        }
    }

    return update;
}

void BelugaSerial::get_ranges(std::vector<BelugaNeighbor> &list) {
    list.clear();
    try {
        list = _range_queue.get(false);
    } catch (const BelugaQueueException &exc) {
        if (exc.reason() != BelugaQueueException::QUEUE_EMPTY) {
            throw;
        }
    }
}

RangeEvent BelugaSerial::get_range_event() {
    RangeEvent event{};
    try {
        event = _range_event_queue.get(false);
    } catch (const BelugaQueueException &exc) {
        if (exc.reason() != BelugaQueueException::QUEUE_EMPTY) {
            throw;
        }
    }
    return event;
}

void BelugaSerial::register_resync_cb(std::function<void()> cb) {
    _time_resync = std::move(cb);
}

void BelugaSerial::swap_port(const std::string &port) {
    fs::path path = port;

    if (!fs::exists(path)) {
        throw std::invalid_argument(port + " is not a valid path");
    }

    if (_tasks_running) {
        throw std::runtime_error(
            "Cannot swap ports if port is still in use! PLease call stop() "
            "before calling this function.");
    }

    _serial.close();
    _serial.port(port);
    _serial.open();
}
} // namespace BelugaSerial
